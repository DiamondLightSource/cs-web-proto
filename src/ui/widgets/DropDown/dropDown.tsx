import React from "react";

import classes from "./dropDown.module.css";
import { Widget } from "../widget";
import { WidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  BoolPropOpt,
  StringProp,
  StringOrNumPropOpt,
  InferWidgetProps,
  ChildrenPropOpt,
  FontPropOpt,
  BorderPropOpt,
  ColorPropOpt
} from "../propTypes";

const DropDownContainerProps = {
  title: StringProp,
  open: BoolPropOpt,
  font: FontPropOpt,
  border: BorderPropOpt,
  minHeight: StringOrNumPropOpt,
  foregroundColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  children: ChildrenPropOpt
};

const DropDownWidgetProps = {
  ...DropDownContainerProps,
  ...WidgetPropType
};

export const DropDownComponent = (
  props: InferWidgetProps<typeof DropDownContainerProps>
): JSX.Element => (
  <details
    className={classes.Detail}
    open={props.open ?? false}
    style={{
      ...props.font?.css(),
      ...props.border?.css(),
      color: props.foregroundColor?.rgbaString() ?? "",
      backgroundColor: props.backgroundColor?.rgbaString() ?? "",
      minHeight: props.minHeight ?? ""
    }}
  >
    <summary className={classes.Summary}>{props.title}</summary>
    <div
      className={classes.Children}
      style={{ position: "relative", height: "100%", width: "100%" }}
    >
      {props.children}
    </div>
  </details>
);

type DropDownWidget = InferWidgetProps<typeof DropDownWidgetProps>;

export const DropDown = (props: DropDownWidget): JSX.Element => (
  <Widget baseWidget={DropDownComponent} {...props} />
);

registerWidget(DropDown, DropDownWidgetProps, "dropdown");
