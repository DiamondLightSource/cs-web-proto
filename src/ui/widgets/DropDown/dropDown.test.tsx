import React from "react";
import { create } from "react-test-renderer";

import { DropDownComponent } from "./dropDown";
import { fireEvent, render, waitFor } from "@testing-library/react";

const grouping = (
  <DropDownComponent title={"Test"}>Test Text</DropDownComponent>
);

describe("<DropDownComponent />", (): void => {
  test("it matches the snapshot", (): void => {
    const snapshot = create(grouping);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("details node opens on click", (): void => {
    const { getByRole, getByText } = render(grouping);
    // Content text there even when closed.
    expect(getByText("Test Text")).toBeInTheDocument();
    const dropDown = getByRole("group");
    expect(dropDown).not.toHaveAttribute("open");
    fireEvent.click(dropDown);
    // Wait for the event loop.
    waitFor(() => expect(dropDown).toHaveAttribute("open"));
  });
});
