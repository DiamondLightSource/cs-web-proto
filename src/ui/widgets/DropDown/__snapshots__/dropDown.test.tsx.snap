// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropDownComponent /> it matches the snapshot 1`] = `
<details
  className="Detail"
  open={false}
  style={
    Object {
      "backgroundColor": "",
      "color": "",
    }
  }
>
  <summary
    className="Summary"
  >
    Test
  </summary>
  <div
    className="Children"
    style={
      Object {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    Test Text
  </div>
</details>
`;
